lib/auth/accountrecovery_test.go

```diff
@@ -111,3 +133,264 @@ func TestRecoveryCodeEventsEmitted(t *testing.T) {
+	require.NoError(t, err)
+	require.Len(t, bytes, RecoveryTokenLenBytes)
+
+	// Test expired token.
+	fakeClock.Advance(defaults.RecoveryStartTokenTTL)
+	_, err = srv.Auth().GetUserToken(ctx, startToken.GetName())
+	require.True(t, trace.IsNotFound(err))
+
+	// Test events emitted.
+	event := mockEmitter.LastEvent()
+	require.Equal(t, event.GetType(), events.RecoveryTokenCreateEvent)
+	require.Equal(t, event.GetCode(), events.RecoveryTokenCreateCode)
+	require.Equal(t, event.(*apievents.UserTokenCreate).Name, u.username)
+
+	// Test with recover type PWD.
+	startToken, err = srv.Auth().StartAccountRecovery(ctx, &proto.StartAccountRecoveryRequest{
+		Username:     u.username,
+		RecoveryCode: []byte(u.recoveryCodes[1]),
+		RecoverType:  types.UserTokenUsage_RECOVER_PWD,
+	})
+	require.NoError(t, err)
+	require.Equal(t, types.UserTokenUsage_RECOVER_PWD, startToken.GetUsage())
+
+	// Test with no recover type.
+	_, err = srv.Auth().StartAccountRecovery(ctx, &proto.StartAccountRecoveryRequest{
+		Username:     u.username,
+		RecoveryCode: []byte(u.recoveryCodes[2]),
+	})
+	require.Error(t, err)
+}
+
+func TestStartAccountRecovery_WithLock(t *testing.T) {
+	srv := newTestTLSServer(t)
+	ctx := context.Background()
+	fakeClock := srv.Clock().(clockwork.FakeClock)
+
+	defaultModules := modules.GetModules()
+	defer modules.SetModules(defaultModules)
+	modules.SetModules(&testWithCloudModules{})
+
+	u, err := createUserWithSecondFactorAndRecoveryCodes(srv, "otp")
+	require.NoError(t, err)
+
+	// Test max failed recovery attempt locks both login and further recovery attempt.
+	for i := 1; i <= defaults.MaxAccountRecoveryAttempts; i++ {
+		_, err = srv.Auth().StartAccountRecovery(ctx, &proto.StartAccountRecoveryRequest{
+			Username: u.username,
+		})
+		require.Error(t, err)
+	}
+
+	user, err := srv.Auth().GetUser(u.username, false)
+	require.NoError(t, err)
+	require.True(t, user.GetStatus().IsLocked)
+	require.False(t, user.GetStatus().LockExpires.IsZero())
+	require.False(t, user.GetStatus().RecoveryAttemptLockExpires.IsZero())
+
+	// Advance time to remove lock and attempts.
+	fakeClock.Advance(defaults.AttemptTTL)
+
+	// Trigger login lock.
+	for i := 1; i <= defaults.MaxLoginAttempts; i++ {
+		_, err = srv.Auth().authenticateUser(ctx, AuthenticateUserRequest{
+			Username: u.username,
+			OTP:      &OTPCreds{},
+		})
+		require.Error(t, err)
+
+		if i == defaults.MaxLoginAttempts {
+			require.True(t, trace.IsAccessDenied(err))
+		}
+	}
+
+	// Test recovery is still allowed after login lock.
+	_, err = srv.Auth().StartAccountRecovery(ctx, &proto.StartAccountRecoveryRequest{
+		Username:     u.username,
+		RecoveryCode: []byte(u.recoveryCodes[0]),
+		RecoverType:  types.UserTokenUsage_RECOVER_2FA,
+	})
+	require.NoError(t, err)
+
+	// Trigger max failed recovery attempts.
+	for i := 1; i <= defaults.MaxAccountRecoveryAttempts; i++ {
+		_, err = srv.Auth().StartAccountRecovery(ctx, &proto.StartAccountRecoveryRequest{
```

Comment:
What is this loop testing that we didn't test the first time we got a recovery lock? Would we lose anything if this was a single call instead?

Reply:
You're right, I re-arranged the test to make more sense.